repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -106,20 +100,34 @@ class Session(interpreter: Interpreter)
     _statements.clear()
   }
 
+  private def changeState(newState: SessionState): Unit = {
+    synchronized {
+      _state = newState
+    }
+    stateChangedCallback(newState)
+  }
+
   private def executeCode(executionCount: Int, code: String): String = synchronized {
-    _state = SessionState.Busy()
+    changeState(SessionState.Busy())
```

Comment:
Can we simplify the Session state, do we really need to differentiate the state of session in such a fine-grained way?

State like "running", "busy" or "idle" are all means session is running correctly, do we really need to differentiate? I don't see any code specifically requires such state like "busy".